Provide fluent API to set up Security JPA identity providers programmatically

michalvavrik opened the pull request:
- portion of https://github.com/quarkusio/quarkus/issues/16728
- allows to configure persistence unit for JPA programmatically
- adds support for named persistence units for Security JPA Reactive

sberyozkin commented on extensions/security-jpa-common/runtime/src/main/java/io/quarkus/security/jpa/SecurityJpa.java:
```diff
@@ -0,0 +1,28 @@
+     *
+     * @return Quarkus Security JPA {@link IdentityProvider}s
+     */
+    static Collection<IdentityProvider<?>> jpa() {
```
May be I'm confused, but I thought `security-jpa` was offering an `IdentityProvider` that can be used with basic or form authentication mechanism ?

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/security/HttpSecurity.java:
```diff
@@ -127,6 +130,19 @@ public interface HttpSecurity {
+     * Registers given {@link HttpAuthenticationMechanism} in addition to all other global authentication mechanisms.
+     *
+     * @param mechanism {@link HttpAuthenticationMechanism}
+     * @param identityProviders {@link IdentityProvider}s that should be used for authentication instead
```
Well, CDI registered augmentor may be used with any mechanism.

I think it would make sense not to mix augmentors and identity providers in such methods.

I'd also consider providing a generic option for adding identity providers without focusing on JPA first, as part of the fluent API, where an option to add an identity provider becomes available only after a method registering a custom mechanism is chosen...

IMHO it should be a draft PR

michalvavrik replied:
> Well, CDI registered augmentor may be used with any mechanism.

Sure. What do you want to say? This PR gives you 3 options:

- use CDI augmentors if you don't configure augmentors via `HttpSecurity`
- use your own augmentors, whether you create instance or inject them
- use your own augmentors and CDI augmentors. Nothing stops you to do `new ArrayList<>()` then `addAll(cdiAugmentors)` and `addAll(customAugmentors)`

sberyozkin replied:
Thanks @michalvavrik, right, so we have `.mechanism(HttpAuthenticationMechanism)` right now, so, perhaps we can, to start with, offer  `.mechanism(HttpAuthenticationMechanism, IdentityProvider)`, `.mechanism(HttpAuthenticationMechanism, IdentityProvider, SecurityIdentityAugmentor)`, that are typical setup options for custom authentication mechanisms ? 

As far as a JPA identity provider is concerned,  you have a good point that it is already CDI registered, but I'm thinking it would a bit confusing to users the way it is offered in this PR, if they say want to add it to a custom mechanism dealing with a name and password, seeing a collection, references to CDI, etc.

Perhaps, when it comes to JPA, it, it should be offered in a typed fashion for existing typed `basic` or `form` methods or as a single `IdentityProvider` produced by a JPA IdentityProvider builder. 

But what should be offered is a JPA setup, like a DB setup (persistence unit name), as the whole idea of the API is to let users avoid `application.properties`.

So for example,

`.basic().jpa().persistence("orders")...`, `.form().jpa().persistence("orders")...`, 

and with the generic method added here, 

`.mechanism(new MyHttpMechanism(), JpaIdenityProvider().persistence("order").build())`
  
Perhaps something like this...

michalvavrik replied:
> `basic(jpa().persistence("abc"))` looks ok too, so the `basic` override will take `IdentityProvider` ? That should be fine and perhaps we can fail early if the identity provider's request type is not `UsernamePasswordAuthenticationRequest`

I think it will take soe other trick, it doesn't make sense to have different `jpa()` static method for `form` and different for `basic` auth mechanism. I am starting working on this right now.

michalvavrik replied:
@sberyozkin I think I know what you mean, but:

1. form-based auth mechanism also needs `IdentityProvider<TrustedAuthenticationRequest>`, so I planned to hide that from users since you are right it doesn't make sense to create `basic(jpa())` with the `IdentityProvider<TrustedAuthenticationRequest>`, but we can't have 2 `jpa()` methods returning 2 different types; ---> I have some proposal for that, I'll show you when I push it. Probably tomorrow.
2. ` jpa().persistence("abc")` is behind why I keep rewriting it again and again, and I am starting to think that this PR was better without `persistence("abc")`. The thing is, the `persistence("abc")` is determined by the package where user definition Hibernate entity is defined in. So, basically, we should be able to infer the persistence unit during the build (Hibernate knows which entity belongs to which PU), because currently, you cannot have 2 entities holding the identity information. I would rather just drop `quarkus.security-jpa.persistence-unit-name` and do the inference for user OOTB (I am almost sure it is possible, because whenever I tried to set different unit, validation failed on the Hibernate side). But that must be handled separately. I'd start with `basic(jpa())` only.

Anyway, thanks for the input, I'll produce something soon (tomorrow?) and then we can discuss it based on the actual code.

michalvavrik replied:
I have opened https://github.com/quarkusio/quarkus/issues/51480.

sberyozkin commented on docs/src/main/asciidoc/security-jpa.adoc:
```diff
@@ -222,6 +222,64 @@ For more information about proactive authentication, see the Quarkus xref:securi
+    void configure(@Observes HttpSecurity httpSecurity) {
+            httpSecurity
+                    .basic(jpa().persistence("basic-pu"))   <1>
+                    .mechanism(Form.create(), jpa().persistence("form-pu"));    <2>
```
I'd restrict to `basic` and `form` only, as with `.mechanism()` there is no identity provider type safety, might work if the custom mechanism is dealing with user name and passwords or might not... I'd let `basic` and `form` users try this option to start with

sberyozkin commented on extensions/security-jpa-common/runtime/src/main/java/io/quarkus/security/jpa/SecurityJpa.java:
```diff
@@ -0,0 +1,89 @@
+ * <ul>
+ * <li>
+ * You want to configure the Basic authentication to use the Quarkus Security JPA {@link IdentityProvider},
+ * while other authentication mechanism is using different {@link IdentityProvider}:
```
I'd drop the follow up, let users try simple things, basic or form with jpa and we can deal with more complex setups later

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/security/HttpSecurity.java:
```diff
@@ -144,6 +145,14 @@ public interface HttpSecurity {
+     * @param identityProviderBuilder such as the Quarkus Security JPA {@link IdentityProviderBuilder}.
+     * @return HttpSecurity
+     */
+    HttpSecurity basic(IdentityProviderBuilder identityProviderBuilder);
```
IMHO it should be a typed `IdentityProvider<UsernamePasswordAuthenticationRequest>`

sberyozkin replied:
You are right there could be more than one, in the architecture, we can later a List as well, I don't really know if anyone uses more than one...

sberyozkin commented on extensions/security/runtime-spi/src/main/java/io/quarkus/security/spi/runtime/IdentityProviderManagerBuilder.java:
```diff
@@ -0,0 +1,23 @@
+ * CDI bean that enables Quarkus core extensions to create own {@link IdentityProviderManager}
+ * from the global {@link IdentityProviderManager} used by Quarkus.
+ */
+public interface IdentityProviderManagerBuilder {
```
This interface introduces a new and rather advanced dimension that IMHO we should not deal with in this PR, I propose to finalize setting typed identity providers starting from JPA and iterate

michalvavrik replied:
I'll try to hind it from users entirely (make it package private class). it is not about providing users, it is about providing this to ourselves (this is how I can choose JPA providers for basic auth for example).

michalvavrik replied:
Well, not sure how I'll hide it. I cannot simply duplicate whole `IdentityProviderManager` logic just to avoid a bean. I'll move it to the deployment build step, it will be much more work, but that will hide this from users. Vertx HTTP only sees SPI, not actual Security extension.

michalvavrik replied:
No, I'll just move it to deployment-spi, that will hide it from users.

michalvavrik replied:
`IdentityProviderManagerBuilder` is removed now.

michalvavrik committed:
feat(security-jpa): configure multiple PUs programmatically

- portion of https://github.com/quarkusio/quarkus/issues/16728
- allows to configure persistence unit for JPA programmatically
- adds support for named persistence units for Security JPA Reactive

Signed-off-by: Michal Vavřík <michal.vavrik@aol.com>

sberyozkin commented on extensions/security-jpa-common/runtime/src/main/java/io/quarkus/security/jpa/SecurityJpa.java:
```diff
@@ -0,0 +1,76 @@
+     * @param persistenceUnitName persistence unit name
+     * @return new {@link IdentityProvider}
+     */
+    static IdentityProvider<UsernamePasswordAuthenticationRequest> jpa(String persistenceUnitName) {
```
@michalvavrik Hi Michal

Right, the point of using this fluent API is to let users not to configure anything related to `security-jpa` in `application.properties`.

My question was more about, do we need to replace the identity provider shipped with `security-jpa` because if we don't, it will just throw some exception because it can't find anything in `application.properties` ?

>If you configure jpa("pu-1") and we are going to apply globally configured jpa()...

What is a globally configured JPA provider without any configuration in `application.properties` ?

I'm just saying that it is ok if some default provider returns null/void uni, the security architecture handles it by delegating to the next one, which will use `pu-1`...
